### PR TITLE
Create base model interface and Firestore model class

### PR DESCRIPTION
**Pull request by josueetcom**
The `FirestoreModel` class does the heavy lifting! The subclasses just need to implement a friendlier contructor.

### CODE REVIEW

File: src/models/FirestoreModel.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+  get(): Promise<FirestoreModel<T>> {
+    return new Promise((resolve, reject) => {
+      this.ref.get()
+        .then((doc: firestore.DocumentSnapshot<T>) => {
+          if (!doc.exists) {
+            reject(new Error(`${this.ref.path} not found`));
+            return;
+          }
+          resolve(new FirestoreModel<T>(doc.ref, doc.data() as T));
```
Comment (edwingl10) [inline src/models/FirestoreModel.ts:19]: `doc.data() as T`
Would this interpret the data as our class models? e.g. User

File: src/models/FirestoreModel.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  delete(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.ref.delete()
+        .then(() => resolve())
+        .catch((error) => reject(error));
+    });
+  }
+
+  subscribe(onChange: (model: Model<T>) => void, onError = console.error): () => void {
```
Comment (edwingl10) [inline src/models/FirestoreModel.ts:41]: I'm having trouble understanding what this does, what happens when there is a change?

File: src/models/FirestoreModel.ts (hunk 1 above)
Comment (josueetcom) [reply src/models/FirestoreModel.ts:19]: Yes that's right

File: src/models/FirestoreModel.ts (hunk 2 above)
Comment (josueetcom) [reply src/models/FirestoreModel.ts:41]: This is a wrapper around [`DocumentReference.onSnapshot`](https://firebase.google.com/docs/reference/js/firebase.firestore.DocumentReference#onsnapshot), which is used to [listen for realtime updates](https://firebase.google.com/docs/firestore/query-data/listen).

Basically what happens is whenever an update to the document happens (e.g. either on the initial call or due to other updates), if the document exists `onChange` gets invoked with a `FirestoreModel` object that contains the data.

Comment (josueetcom) [reply src/models/FirestoreModel.ts:41]: [Here](https://github.com/beyondhb1079/s4us/blob/95bd0c1a7d51a05d40ec66ebceae1226edd4d697/src/pages/ScholarshipDetails.jsx#L12-L17)'s how it's used in the follow-up PR

Comment (edwingl10) [reply src/models/FirestoreModel.ts:41]: Ohh I see, that makes sense!

**Review by edwingl10 — APPROVED** [review]

File: src/models/FirestoreModel.ts (hunk 2 above)
Comment (josueetcom) [reply src/models/FirestoreModel.ts:41]: Yep! One more note. The returned function is a function to detach the listener, in this case it's just the function returned by `DocumentReference.onSnapshot`. That detach function should be invoked whenever a component unmounts and/or needs to be updated with a `onChange` listener.

**Review by edwingl10 — APPROVED** [review]

**Pull request merged**